### PR TITLE
Adding ``coupling-map`` and ``target`` to ``HighLevelSynthesis`` 

### PR DESCRIPTION
**Pull request by alexanderivrii**
### Summary

This PR adds the arguments ``coupling_map``, ``target`` and ``use_qubit_indices`` to ``HighLevelSynthesis`` transpiler pass, and adds the arguments ``coupling_map``, ``targets`` and ``qubits`` to ``HighLevelSynthesisPlugin``. The change is backward-compatible.

### Details and comments

This reimplements yet another component of #9250, and consists of defining a cleaner API for `HighLevelSynthesis` and `HighLevelSynthesisPlugin` that takes `coupling_map` and `target` into account. See discussion [here](https://github.com/Qiskit/qiskit-terra/pull/9250#issuecomment-1497404165) and [here](https://github.com/Qiskit/qiskit-terra/pull/9250#issuecomment-1632341694).

The argument ``use_qubit_indices`` provides the context to ``HighLevelSynthesis`` whether it's running before or after the layout has been set, that is, whether the qubit indices of higher-level-objects correspond to qubit indices on the target backend. One thing which I did not realize initially is that even when passes like unroll3q or high-level-synthesis are running on the abstract circuit, they have access to `target` and hence to the coupling map of the device (this was not this way before `target` was added).

### CODE REVIEW

File: qiskit/transpiler/passes/synthesis/high_level_synthesis.py
Comment (mtreinish) [inline qiskit/transpiler/passes/synthesis/high_level_synthesis.py:218]: There's actually a new interface for this added in 0.25, you can do:

```suggestion
                    qubits=[dag.find(x).index for x in node.qargs],
```

File: qiskit/transpiler/passes/synthesis/high_level_synthesis.py
Comment (mtreinish) [inline qiskit/transpiler/passes/synthesis/high_level_synthesis.py:167]: Paired with the suggestion below at L218 you can skip this step:

```suggestion
```

File: releasenotes/notes/reno-691a6036754487df.yaml
Comment (mtreinish) [inline releasenotes/notes/reno-691a6036754487df.yaml:8]: I think having some examples here would go a long way for plugin maintainers.

File: test/python/transpiler/test_high_level_synthesis.py
Comment (mtreinish) [inline test/python/transpiler/test_high_level_synthesis.py:127]: I guess it doesn't matter in this case since we're just testing the raise, but  you can do `target.build_coupling_map()` to get a coupling map from a target.

File: qiskit/transpiler/preset_passmanagers/common.py
Comment (mtreinish) [inline qiskit/transpiler/preset_passmanagers/common.py:220]: I'm wondering if there is a potential issue here because unroll3q is typically run before layout so the qubit indices that get set in the pass will not be indices in the coupling map (ie qubit 0 in the circuit is not yet decided to be qubit 0 on the target backend). Maybe we need a flag to skip the qubits or otherwise flag that we've yet to set a layout.

File: releasenotes/notes/reno-691a6036754487df.yaml
Comment (mtreinish) [inline releasenotes/notes/reno-691a6036754487df.yaml:8]: Also it might be worthwhile to call out that if they're not explicitly added to a plugin's `run()` method kwargs the coupling map, target, and qubits will end up in the options input.

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM, the interface makes sense as it's explicitly setting the required arguments. Just a few small questions/comments inline. One other thing is do you think we should add some details about this to the plugin interface docs: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/passes/synthesis/plugin.py#L159 because we're adding on to the interface.

File: qiskit/transpiler/passes/synthesis/high_level_synthesis.py
Comment (alexanderivrii) [reply qiskit/transpiler/passes/synthesis/high_level_synthesis.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/10477/commits/11f92363af5b536a0b0ce3444e4e2795b70419e9. But the method is called `find_bit`, not `find` :p (this is supposed to be sticking-tongue-out-emoji).

File: qiskit/transpiler/passes/synthesis/high_level_synthesis.py
Comment (alexanderivrii) [reply qiskit/transpiler/passes/synthesis/high_level_synthesis.py]: Done in https://github.com/Qiskit/qiskit-terra/commit/11f92363af5b536a0b0ce3444e4e2795b70419e9.

File: qiskit/transpiler/passes/synthesis/high_level_synthesis.py
Comment (mtreinish) [reply qiskit/transpiler/passes/synthesis/high_level_synthesis.py]: Oops my bad, that's what I get for reviewing at the end of the day. I didn't notice my own typos :)

File: releasenotes/notes/reno-691a6036754487df.yaml
Comment (alexanderivrii) [reply releasenotes/notes/reno-691a6036754487df.yaml]: I agree. At the moment I don't have any good examples, but I would after the last part of #9250 gets cleaned up, i.e. actually having some plugins that take `coupling_map` and `qubits` into account. And then we would be able to mention/reference these specific plugins in the plugin interface docs. By the way, thanks for reminding to update these docs, that's done in [54bd00b](https://github.com/Qiskit/qiskit-terra/pull/10477/commits/54bd00b9cdb66e93b81d664b8c8ca433cd6d6e34).

File: test/python/transpiler/test_high_level_synthesis.py
Comment (alexanderivrii) [reply test/python/transpiler/test_high_level_synthesis.py:127]: Yes, good to know, thanks!

File: qiskit/transpiler/preset_passmanagers/common.py
Comment (alexanderivrii) [reply qiskit/transpiler/preset_passmanagers/common.py:220]: That's a good point. At the moment I think having `qubits=None` is exactly an indication that the layout has not yet been chosen. I have added a sentence on that to the plugin reference docs.

File: releasenotes/notes/reno-691a6036754487df.yaml
Comment (alexanderivrii) [reply releasenotes/notes/reno-691a6036754487df.yaml]: I made a pass over release notes in https://github.com/Qiskit/qiskit-terra/pull/10477/commits/5a1365a7fdb8108f39430e317526d9e1d5307e2e.

File: qiskit/transpiler/preset_passmanagers/common.py
Comment (mtreinish) [reply qiskit/transpiler/preset_passmanagers/common.py:220]: But won't this still set `qubits` to a value because it's unconditionally populated in the `plugin_method.run()` call?

Comment (alexanderivrii) [reply qiskit/transpiler/preset_passmanagers/common.py]: Update: I spoke with Matthew in private, and he has convinced me that `HighLevelSynthesis` needs to know whether it's running before or after the layout is set, and suggested to add a flag `use_qubit_indices` to `HighLevelSynthesis` transpiler pass. This is now done in https://github.com/Qiskit/qiskit-terra/pull/10477/commits/e3211fbe69dcd93c471ca1de09157b965ccc3cbf.

File: test/python/transpiler/test_high_level_synthesis.py
Comment (alexanderivrii) [reply test/python/transpiler/test_high_level_synthesis.py:128]: I have also added a test that when only the target is set, the synthesis plugin does see a valid coupling map (because indeed the `HighLevelSynthesis` pass builds it using the function above).

File: qiskit/transpiler/passes/synthesis/plugin.py
Comment (mtreinish) [inline qiskit/transpiler/passes/synthesis/plugin.py:180]: Since you've defined defaults for the arguments they're typically labelled as keyword arguments instead of positional. Although in Python they can be done either way. 

```suggestion
The keyword argument ``target`` specifies the target backend, allowing the plugin
to access all target-specific information,
such as the coupling map, the supported gate set, and so on. The keyword argument
``coupling_map`` only specifies the coupling map, and is only used when ``target``
is not specified.
The keyword argument ``qubits`` specifies the list of qubits over which the
higher-level-object is defined, in case the synthesis is done on the physical circuit.
```

File: qiskit/transpiler/passes/synthesis/plugin.py
Comment (mtreinish) [inline qiskit/transpiler/passes/synthesis/plugin.py:181]:
```suggestion
The value of ``None`` indicates that the layout has not yet been chosen and the physical qubits
in the target or coupling map that this operation is operating on has not yet been determined.
```

File: qiskit/transpiler/preset_passmanagers/common.py
Comment (mtreinish) [inline qiskit/transpiler/preset_passmanagers/common.py:433]: We can set this `True` here I think because we've applied a layout and routed the circuit, so if there are any higher level objects to synthesize here, their qargs are physical qubits in the target/coupling map.

```suggestion
                use_qubit_indices=True,
```

File: qiskit/transpiler/preset_passmanagers/common.py
Comment (mtreinish) [inline qiskit/transpiler/preset_passmanagers/common.py:456]:
```suggestion
                use_qubit_indices=True,
```

File: qiskit/transpiler/preset_passmanagers/common.py
Comment (mtreinish) [inline qiskit/transpiler/preset_passmanagers/common.py:477]:
```suggestion
                use_qubit_indices=True,
```

**Review by mtreinish — COMMENTED** [review]
This LGTM, thanks for the quick update. I had a couple of small suggestions inline but nothing worth really blocking over. If you don't think we should make these changes I'm happy to approve and move forward with this.

File: qiskit/transpiler/preset_passmanagers/common.py
Comment (mtreinish) [inline qiskit/transpiler/preset_passmanagers/common.py:222]: This should be safe to do now that we have `use_qubit_indices=False`. This enables you to do something like find the longest path for LNN synthesis before layout/routing.

```suggestion
            hls_config=hls_config, coupling_map=coupling_map, target=target, use_qubit_indices=False
```

**Review by mtreinish — DISMISSED** [review]
LGTM, thanks for the quick update.

**Review by mtreinish — APPROVED** [review]

**Comment by alexanderivrii** [comment]
@mtreinish, many thanks for the improvements! Though the commit [af27ba8](https://github.com/Qiskit/qiskit-terra/pull/10477/commits/af27ba8aade6d0fa18de41b26a23d0db1521c48e) was problematic: the function `generate_unroll_3q` does not have the argument `coupling_map`. However, it does accept `target` that gets passed to `HighLevelSynthesis`, allowing it to build the coupling map from the target. And this is indeed very nice -- exactly as you say this allows us to choose a better method for synthesizing a high-level-object before layout/routing,

**Review by mtreinish — APPROVED** [review]
Thanks, you're correct, I neglected that the `coupling_map` isn't available there. Sorry for applying an incorrect suggestion. This is ready to go from my perspective then.

**Comment by jakelishman** [comment]
Sorry, I just removed this from the queue because I think it may be worth holding off merging anything until the cross-repo merge I'm just writing up now is complete.

**Pull request merged**
